## Part 1

In [1]:
import numpy as np # for working with vectorized data
import pandas as pd # for data analysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import requests # needed to retrive html from a web address
from bs4 import BeautifulSoup # for parsing html so data can be extracted

The first step is to scrape the relevent data and store it in a parsable object

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M' # toronto postcodes begin with M and the table has all desired data so joins are not needed to complete the dataframe
html_data = requests.get(url).text # saves html as a single string
soup = BeautifulSoup(html_data, 'html5lib') # creates a BeautifulSoup object which parses the html_data in the same way as a browser

The next step is to create an itterable of the elements each containing a single record

In [3]:
table = soup.tbody.find_all('td') # finds the first tbody element and creates an iterable of td elements within (cells from the table)

The records will now be iterated through, cleaned, and reorganised into a useful format

In [4]:
data = {'PostalCode':[], 'Borough':[], 'Neighborhood':[]} # creates a dictionary of empty lists. the keys will become column titles
for cell in table: # iterates through cells in the table taken from wikipedia
    
    if cell.span.text == 'Not assigned': # checks if the post code is not in use and skips if True
        continue
    
    else:
        data['PostalCode'].append(cell.text[1:4]) # slices postcode from cell because length and format are consistent and appends to the 'PostalCode' list in data
        
        borough, neighborhood = cell.text[4:].split('(', 1) # splits all text after the postcode at the first bracket (how borough and neighborhood are seperated on the website table)
        
        data['Borough'].append(borough) # borough is already cleaned and gets appended to the 'Borough' list in data
        
        neighborhood = neighborhood.strip('\n()').replace(' /', ',') # before neighborhood can be appended to its list all remaining brackets and newlines are stripped and replace is used to change / seperators into comma seperators
        
        # checks if the neighborhood(s) is listed for the borough and lables as the borough if not, although the table doesn't actually have this issue currently so it is commented out
        #if neighborhood == 'Not assigned': 
            #neighborhood = borough
        #else:
            #pass
            
        data['Neighborhood'].append(neighborhood) 

data

{'PostalCode': ['M3A',
  'M4A',
  'M5A',
  'M6A',
  'M7A',
  'M9A',
  'M1B',
  'M3B',
  'M4B',
  'M5B',
  'M6B',
  'M9B',
  'M1C',
  'M3C',
  'M4C',
  'M5C',
  'M6C',
  'M9C',
  'M1E',
  'M4E',
  'M5E',
  'M6E',
  'M1G',
  'M4G',
  'M5G',
  'M6G',
  'M1H',
  'M2H',
  'M3H',
  'M4H',
  'M5H',
  'M6H',
  'M1J',
  'M2J',
  'M3J',
  'M4J',
  'M5J',
  'M6J',
  'M1K',
  'M2K',
  'M3K',
  'M4K',
  'M5K',
  'M6K',
  'M1L',
  'M2L',
  'M3L',
  'M4L',
  'M5L',
  'M6L',
  'M9L',
  'M1M',
  'M2M',
  'M3M',
  'M4M',
  'M5M',
  'M6M',
  'M9M',
  'M1N',
  'M2N',
  'M3N',
  'M4N',
  'M5N',
  'M6N',
  'M9N',
  'M1P',
  'M2P',
  'M4P',
  'M5P',
  'M6P',
  'M9P',
  'M1R',
  'M2R',
  'M4R',
  'M5R',
  'M6R',
  'M7R',
  'M9R',
  'M1S',
  'M4S',
  'M5S',
  'M6S',
  'M1T',
  'M4T',
  'M5T',
  'M1V',
  'M4V',
  'M5V',
  'M8V',
  'M9V',
  'M1W',
  'M4W',
  'M5W',
  'M8W',
  'M9W',
  'M1X',
  'M4X',
  'M5X',
  'M8X',
  'M4Y',
  'M7Y',
  'M8Y',
  'M8Z'],
 'Borough': ['North York',
  'North York',
  'Downtown Tor

Now that the data is appropriatly formated as a dictionary of equal length lists it can be converted into a pandas dataframe

In [5]:
df = pd.DataFrame(data)
df

PostalCode                                            Borough  \
0          M3A                                         North York   
1          M4A                                         North York   
2          M5A                                   Downtown Toronto   
3          M6A                                         North York   
4          M7A                                       Queen's Park   
5          M9A                                          Etobicoke   
6          M1B                                        Scarborough   
7          M3B                                         North York   
8          M4B                                          East York   
9          M5B                                   Downtown Toronto   
10         M6B                                         North York   
11         M9B                                          Etobicoke   
12         M1C                                        Scarborough   
13         M3C                                         North York   
14         M4C                                          East York   
15         M5C                                   Downtown Toronto   
16         M6C                                               York   
17         M9C                                          Etobicoke   
18         M1E                                        Scarborough   
19         M4E                                       East Toronto   
20         M5E                                   Downtown Toronto   
21         M6E                                               York   
22         M1G                                        Scarborough   
23         M4G                                          East York   
24         M5G                                   Downtown Toronto   
25         M6G                                   Downtown Toronto   
26         M1H                                        Scarborough   
27         M2H                                         North York   
28         M3H                                         North York   
29         M4H                                          East York   
30         M5H                                   Downtown Toronto   
31         M6H                                       West Toronto   
32         M1J                                        Scarborough   
33         M2J                                         North York   
34         M3J                                         North York   
35         M4J                              East YorkEast Toronto   
36         M5J                                   Downtown Toronto   
37         M6J                                       West Toronto   
38         M1K                                        Scarborough   
39         M2K                                         North York   
40         M3K                                         North York   
41         M4K                                       East Toronto   
42         M5K                                   Downtown Toronto   
43         M6K                                       West Toronto   
44         M1L                                        Scarborough   
45         M2L                                         North York   
46         M3L                                         North York   
47         M4L                                       East Toronto   
48         M5L                                   Downtown Toronto   
49         M6L                                         North York   
50         M9L                                         North York   
51         M1M                                        Scarborough   
52         M2M                                         North York   
53         M3M                                         North York   
54         M4M                                       East Toronto   
55         M5M                                         North York   
56         M6M                                               York   
57

In [6]:
df.shape

(103, 3)

## Part 2

Now that we have collected 103 postcodes with their list of neighborhoods, we can find the corresponding latitudes and longitudes

In [7]:
#!pip install geocoder # uncomment if geocoder is not installed
import geocoder # for converting addresses into latitude and longitude
from geopy.geocoders import Nominatim # an alternitive api library to geocoder in case it doesn't work

```python
# couldn't get the geocoder library to work

latitude = []

longitude = []

for post_code in df['PostalCode']:
    
    lat_lng_coords = None # initialize variable to None
    
    while(lat_lng_coords is None): # loop until the server responds
    
        g = geocoder.osm('{}, Toronto, Ontario'.format(post_code))
        lat_lng_coords = g.osm
        
    latitude.append(lat_lng_coords['x'])
    longitude.append(lat_lng_coords['y'])

df['Latitude'] = latitude
df['Longitude'] = longitude
```

```python
# other library also timed out
geolocator = Nominatim(user_agent="scott_pilgrim")
latitude = []
longitude = []

for address in df['PostalCode']:
    location = None
    while(location == None):
        location = geolocator.geocode('{}, Toronto, Ontario'.format(address))

    print('{} address got'.format(address))
    latitude.append(location.latitude)
    longitude.append(location.longitude)

df['Latitude'] = latitude
df['Longitude'] = longitude
```

In [8]:
# had to resort to using a premade csv
lat_lng_df = pd.read_csv('Geospatial_Coordinates.csv')
lat_lng_df

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

Downloding a full dataframe like this requires another step to combine the previous dataframe

In [9]:
df = pd.merge(df, lat_lng_df, left_on='PostalCode', right_on='Postal Code') # uses post code as a key to align the tables
df = df.drop('Postal Code', axis=1)

## Part 3

In [10]:
address = 'Toronto, ON' # create search string. this doesn't need a set format but performs better on comma seperated key words

geolocator = Nominatim(user_agent="canada_eh") # creates an object out of the geocoder service's API, which abstracts its functions
location = geolocator.geocode(address) # the geocode method turns a query string into a dictionary like geopy.location.Location object for any geocoder
latitude = location.latitude # latitude and longitude are keys
longitude = location.longitude
print('Toronto is located at {}, {}.'.format(latitude, longitude))

Toronto is located at 43.6534817, -79.3839347.


In [11]:
# fixed argumets for Foursquare api
CLIENT_ID = 'IESGEVMGJBKRSU3C4PVB2HKV5DSTHLKH1DSQKYS2WQ4JHPP2' # Foursquare ID
CLIENT_SECRET = 'EPYE2PPFYKLAR23NB50ZODOAFKOJCW5DBISX25DKWXORZHFP' # Foursquare Secret
VERSION = '20210623' # Foursquare API version YYYY/MM/DD = 2021/06/23
RADIUS = 500 # meters away that venues can be from the coordinates given
LIMIT = 100 # the maximum number of items the API will return

Here I will collect information on boroughs rather than individule neighborhoods to limit api calls

In [12]:
toronto_df = df.copy(deep=True) # creates a copy so this section can be done seperately

# iterates through each row
for i in range(toronto_df.shape[0]):
    
    borough_latitude = toronto_df.loc[i, 'Latitude'] # borough latitude value
    borough_longitude = toronto_df.loc[i, 'Longitude'] # borough longitude value

    # creates an api call to be used for each borough coordinates
    url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID,CLIENT_SECRET,borough_latitude,borough_longitude,VERSION,RADIUS,LIMIT)
    # creates a GET request and filters it to just the relevent list
    results = requests.get(url).json()["response"]['groups'][0]['items']
        
    # extracts all venue types and creates a column for each which counts the number within 500m of the borough's coordinates
    for v in results:
        
        # filter applied to each item the api returns
        venue_type = v['venue']['categories'][0]['name']
        
        #  skips neighborhood venue type since it is uninformative
        if venue_type == 'Neighborhood':
            continue
        
        # if the venue types has not appeared in the data previously this will make a column of zero counts on each row
        elif venue_type not in toronto_df.columns:
            
            toronto_df[venue_type] = 0
            
        # itterates the relevent cell by one for each occurence 
        toronto_df.loc[i, venue_type] += 1
    
toronto_df

PostalCode                                            Borough  \
0          M3A                                         North York   
1          M4A                                         North York   
2          M5A                                   Downtown Toronto   
3          M6A                                         North York   
4          M7A                                       Queen's Park   
5          M9A                                          Etobicoke   
6          M1B                                        Scarborough   
7          M3B                                         North York   
8          M4B                                          East York   
9          M5B                                   Downtown Toronto   
10         M6B                                         North York   
11         M9B                                          Etobicoke   
12         M1C                                        Scarborough   
13         M3C                                         North York   
14         M4C                                          East York   
15         M5C                                   Downtown Toronto   
16         M6C                                               York   
17         M9C                                          Etobicoke   
18         M1E                                        Scarborough   
19         M4E                                       East Toronto   
20         M5E                                   Downtown Toronto   
21         M6E                                               York   
22         M1G                                        Scarborough   
23         M4G                                          East York   
24         M5G                                   Downtown Toronto   
25         M6G                                   Downtown Toronto   
26         M1H                                        Scarborough   
27         M2H                                         North York   
28         M3H                                         North York   
29         M4H                                          East York   
30         M5H                                   Downtown Toronto   
31         M6H                                       West Toronto   
32         M1J                                        Scarborough   
33         M2J                                         North York   
34         M3J                                         North York   
35         M4J                              East YorkEast Toronto   
36         M5J                                   Downtown Toronto   
37         M6J                                       West Toronto   
38         M1K                                        Scarborough   
39         M2K                                         North York   
40         M3K                                         North York   
41         M4K                                       East Toronto   
42         M5K                                   Downtown Toronto   
43         M6K                                       West Toronto   
44         M1L                                        Scarborough   
45         M2L                                         North York   
46         M3L                                         North York   
47         M4L                                       East Toronto   
48         M5L                                   Downtown Toronto   
49         M6L                                         North York   
50         M9L                                         North York   
51         M1M                                        Scarborough   
52         M2M                                         North York   
53         M3M                                         North York   
54         M4M                                       East Toronto   
55         M5M                                         North York   
56         M6M                                               York   
57

This code was used to determin "can only concatenate str (not "int") to str" errors were caused by neighborhood being a venue type in the raw data
```python
try:
    toronto_df.loc[i, venue_type] += 1
            
except: 
    print('The title is {}, the row is {}, the dtype is {}, and...'.format(venue_type, i, toronto_df[venue_type].dtypes))
    print('This is the venue:', v)
```

Now there is a glut of data to build a clustering model

In [24]:
from sklearn.preprocessing import StandardScaler # will be used to normalize the dataset
from sklearn.cluster import KMeans # the k means model builder

centres = 5 # the number of clusters the data will be split into

X = toronto_df.values[:,5:] # the features to be normalized
cluster_dataset = StandardScaler().fit_transform(X)
print("normalized data: ", cluster_dataset)
k_means = KMeans(init="k-means++", n_clusters=centres, n_init=20)
k_means.fit(cluster_dataset)
labels = k_means.labels_

df["Labels"] = labels # the catagories are appended to the original dataframe
df.head()

normalized data:  [[ 1.21603891  0.78716867  0.61864129 ... -0.09901475 -0.09901475
  -0.09901475]
 [-0.43201382 -0.71428268 -0.30483774 ... -0.09901475 -0.09901475
  -0.09901475]
 [-0.43201382  3.79007136  0.61864129 ... -0.09901475 -0.09901475
  -0.09901475]
 ...
 [ 1.21603891  0.78716867 -0.30483774 ... -0.09901475 -0.09901475
  -0.09901475]
 [-0.43201382 -0.71428268 -0.30483774 ... -0.09901475 -0.09901475
  -0.09901475]
 [-0.43201382 -0.71428268 -0.30483774 ... 10.09950494 10.09950494
  10.09950494]]


PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Labels  
0 -79.329656       0  
1 -79.315572       0  
2 -79.360636       0  
3 -79.464763       0  
4 -79.389494       0

In [14]:
import folium # module for rendering a map

# cm stands for color map. 
# it takes an array of floats, maps them to positions on a given spectrum and returns arrays of floats denoting intensity of:
# Red, Green, Blue, and alpha (RGBa)
import matplotlib.cm as cm

# This module includes functions and classes for color specification conversions
# will be used to convert RGBa float tuples into  hex codes, which are in ubiquitus use for rendering colour (can also convert RGB, html names, or a single float for greyscale)
import matplotlib.colors as colors 

In [25]:
# set colour scheme for the clusters
colors_array = cm.rainbow(np.linspace(0, 1, centres)) # creates evenly spaced values on the 'rainbow' spectrum
rainbow = [colors.rgb2hex(i) for i in colors_array]  # converts RGBa array to hex code array

In [34]:
# will create a map centered on Toronto
toronto_map = folium.Map(location=[latitude + 0.07, longitude], zoom_start=11) # adjusted latitude for better view of whole city

# adds markers to the map for each borough
for lat, lon, name, cluster in zip(df['Latitude'], df['Longitude'], df['Neighborhood'], df['Labels']):
    label = folium.Popup(str(name) + ' assigned to Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius = 5,
        popup = label,
        color = rainbow[cluster],
        fill = True,
        fill_color = rainbow[cluster],
        fill_opacity = 0.9).add_to(toronto_map)
       
toronto_map